File: frontend/src/adminActivity.js
import { API } from './apiConfig';

let active = null;
const pagePattern = /^\/admin\/(accounts|proxies|addresses|bank-cards|users|wallet|orders|packages|notices|audit|proxy-activity|payment-exceptions)(\/[0-9]+)?$/;
export const adminPage = () => pagePattern.test(location.pathname) ? location.pathname : '';
export const adminAuditHeaders = () => adminPage() ? { 'X-Aitok-Page': adminPage() } : {};
export const activityContext = () => active && adminPage() ? { ...active, page: adminPage() } : null;

// 只发送动作枚举，绝不采集输入框、选项值、按钮原文、查询参数或剪贴板。
export async function recordAdminActivity(context, event) {
  if (!context) return;
  const body = JSON.stringify({ request_key: crypto.randomUUID(), page: context.page, ...event });
  for (let attempt = 0; attempt < 2; attempt++) {
    try {
      const response = await fetch(API + '/admin-activity', {
        method: 'POST', keepalive: true, signal: AbortSignal.timeout(5000),
        headers: { Authorization: `Bearer ${context.token}`, 'Content-Type': 'application/json' }, body,
      });
      if (response.ok || response.status < 500) return;
    } catch {}
    if (!active || active.token !== context.token) return;
  }
  // 不缓存到本地存储，避免跨用户重放；失败也不影响原业务操作。
  console.warn('操作审计暂时无法保存');
}

function controlFor(element) {
  if (element.matches('.account-owner-bind')) return 'bind_user';
  if (element.matches('.account-renewal-switch')) return 'subscription';
  if (element.closest('.pagination-pages')) return 'paginate';
  if (element.getAttribute('role') === 'option' || element.getAttribute('role') === 'combobox') return 'select';
  if (element.matches('a')) return 'link';
  const text = (element.textContent || '').trim();
  const labels = [
    [/^(添加|新增|创建|新建)/, 'add'], [/^编辑/, 'edit'], [/^删除/, 'delete'], [/^确认/, 'confirm'],
    [/^(取消|返回修改|收起|关闭$)/, 'cancel'], [/^保存/, 'save'], [/^(搜索|查找)/, 'search'], [/^(刷新|重试|立即重试)/, 'refresh'],
    [/^导出/, 'export'], [/^(导入|批量导入|测试并导入)/, 'import'], [/^(打开账号|打开浏览器)/, 'open_browser'],
    [/^(关闭浏览器|关闭账号窗口)/, 'close_browser'], [/^(详情|完整资料|查看|使用记录|全部使用记录|余额 \/ 对账单|运营 \/ 核对)/, 'view'],
    [/^复制/, 'copy'], [/^(自动|白色|黑色)$/, 'theme'], [/^(上一页|下一页)$/, 'paginate'], [/^测试/, 'test'],
    [/^获取 IP/, 'get_ip'], [/^分配/, 'assign'], [/^核验/, 'verify'], [/^(退款|申请原路退款|批准原路退款|原申请重试)/, 'refund'],
    [/^记录存入/, 'deposit'], [/^(官网扣款|记录开通扣款|核对并记账)/, 'purchase'], [/^管理分组/, 'group'], [/^废弃/, 'discard'],
    [/^(开启|关闭)续费提醒/, 'subscription'],
  ];
  return labels.find(([pattern]) => pattern.test(text))?.[1] || 'button';
}

export function installAdminActivity(token, userID) {
  const session = { token, userID };
  active = session;
  const click = event => {
    const element = event.target instanceof Element ? event.target.closest('button,a,[role="option"],summary') : null;
    if (!element || element.matches(':disabled,[aria-disabled="true"]')) return;
    // 页面跳转由 page_view 记录一次；其他链接不上传目标 URL。
    if (element.matches('a') && pagePattern.test(element.getAttribute('href') || '')) return;
    void recordAdminActivity(activityContext(), { kind: 'click', control: controlFor(element), result: 'triggered' });
  };
  const submit = () => { void recordAdminActivity(activityContext(), { kind: 'submit', control: 'form', result: 'triggered' }); };
  const change = event => {
    if (event.target.matches('input[type="checkbox"],input[type="radio"],select')) {
      void recordAdminActivity(activityContext(), { kind: 'change', control: 'toggle', result: 'triggered' });
    }
  };
  document.addEventListener('click', click, true);
  document.addEventListener('submit', submit, true);
  document.addEventListener('change', change, true);
  return () => {
    document.removeEventListener('click', click, true);
    document.removeEventListener('submit', submit, true);
    document.removeEventListener('change', change, true);
    if (active === session) active = null;
  };
}

export function localActivityControl(path, method) {
  if (/^\/browsers\/[^/]+\/fingerprint$/.test(path) && method === 'POST') return 'browser_fingerprint';
  if (/^\/browsers\/[^/]+\/proxy$/.test(path) && method === 'PATCH') return 'proxy_bind';
  if (path === '/browsers' && method === 'POST') return 'open_browser';
  if (/^\/browsers\/[^/]+$/.test(path) && method === 'DELETE') return 'close_browser';
  if (path === '/proxies/parse' && method === 'POST') return 'proxy_import';
  if (/^\/proxies(?:\/[^/]+)?\/test$/.test(path) && method === 'POST') return 'proxy_test';
  if (path === '/proxies' && method === 'POST') return 'proxy_create';
  if (/^\/proxies\/[^/]+$/.test(path)) return { GET: 'proxy_read', PATCH: 'proxy_update', DELETE: 'proxy_delete' }[method];
  return null;
}

File: backend/cmd/server/admin_activity.go
package main

import (
	"context"
	"database/sql"
	"encoding/json"
	"io"
	"log"
	"net/http"
	"regexp"
	"strconv"
	"strings"
	"time"
)

var adminPagePattern = regexp.MustCompile(`^/admin/(accounts|proxies|addresses|bank-cards|users|wallet|orders|packages|notices|audit|proxy-activity|payment-exceptions)(/[0-9]+)?$`)
var auditControls = map[string]string{
	"browser_fingerprint": "重新生成浏览器指纹",
	"button":              "点击按钮", "link": "打开链接", "form": "提交表单", "select": "选择选项", "toggle": "切换选项",
	"add": "新增", "edit": "编辑", "delete": "删除", "confirm": "确认操作", "cancel": "取消操作", "save": "保存",
	"search": "搜索", "refresh": "刷新", "export": "导出", "import": "导入", "open_browser": "打开浏览器", "close_browser": "关闭浏览器",
	"view": "查看详情", "copy": "复制", "theme": "切换主题", "paginate": "翻页", "test": "测试代理", "get_ip": "获取出口 IP",
	"assign": "分配处理人", "verify": "核验开通", "refund": "退款", "deposit": "记录存入", "purchase": "记录官网扣款", "group": "管理分组",
	"subscription": "修改是否续订", "proxy_create": "新增本机代理", "proxy_update": "编辑本机代理", "proxy_delete": "删除本机代理",
	"discard":    "废弃订单",
	"bind_user":  "绑定账号所属用户",
	"proxy_read": "查看本机代理", "proxy_test": "测试本机代理", "proxy_import": "解析代理导入", "proxy_bind": "绑定账号代理",
}

type adminActivity struct {
	Key     string `json:"request_key"`
	Page    string `json:"page"`
	Kind    string `json:"kind"`
	Control string `json:"control"`
	Result  string `json:"result"`
}

// 客户端只允许提交固定动作枚举；身份、时间及结果来源由服务端确定。
// 不接受表单值、按钮原文、查询参数、请求正文或客户端指定的操作者。
func (s *Server) adminActivity(w http.ResponseWriter, r *http.Request) {
	if r.Method != "POST" {
		w.WriteHeader(405)
		return
	}
	user, err := s.auth(r)
	if err != nil {
		w.WriteHeader(401)
		return
	}
	r.Body = http.MaxBytesReader(w, r.Body, 2048)
	var in adminActivity
	decoder := json.NewDecoder(r.Body)
	decoder.DisallowUnknownFields()
	if decoder.Decode(&in) != nil || decoder.Decode(&struct{}{}) != io.EOF || !ledgerKeyPattern.MatchString(in.Key) || !adminPagePattern.MatchString(in.Page) {
		w.WriteHeader(400)
		return
	}
	action := auditControls[in.Control]
	switch in.Kind {
	case "page_view":
		if in.Control != "" || in.Result != "visited" {
			w.WriteHeader(400)
			return
		}
		action = "访问页面"
	case "click", "submit", "change":
		if action == "" || in.Result != "triggered" {
			w.WriteHeader(400)
			return
		}
	case "local_request":
		if action == "" || !(strings.HasPrefix(in.Control, "proxy_") || in.Control == "open_browser" || in.Control == "close_browser" || in.Control == "browser_fingerprint") || (in.Result != "success" && in.Result != "failure" && in.Result != "cancelled") {
			w.WriteHeader(400)
			return
		}
	default:
		w.WriteHeader(400)
		return
	}
	if !s.allowAttempt(r.Context(), "admin-activity:"+strconv.FormatInt(user, 10), 300, time.Minute) {
		w.Header().Set("Retry-After", "60")
		w.WriteHeader(429)
		return
	}
	data, _ := json.Marshal(map[string]any{"source": "browser", "page": in.Page, "kind": in.Kind, "control": in.Control, "result": in.Result})
	var id int64
	err = s.db.QueryRowContext(r.Context(), `INSERT INTO operation_events(actor_id,entity_type,entity_id,action,request_key,after_data) VALUES($1,'admin_ui',$1,$2,$3,$4) ON CONFLICT(entity_type,entity_id,request_key) DO NOTHING RETURNING id`, user, action, in.Key, data).Scan(&id)
	if err == sql.ErrNoRows {
		var same bool
		err = s.db.QueryRowContext(r.Context(), `SELECT after_data=$3::jsonb FROM operation_events WHERE entity_type='admin_ui' AND entity_id=$1 AND request_key=$2`, user, in.Key, data).Scan(&same)
		if err == nil && !same {
			reply(w, map[string]string{"error": "审计请求标识已被其他操作使用"}, 409)
			return
		}
	}
	if err != nil {
		reply(w, map[string]string{"error": "操作记录保存失败"}, 503)
		return
	}
	w.WriteHeader(204)
}

type requestAuditContext struct {
	Page, Method, Resource string
	Recorded               bool
}
type requestAuditKey struct{}

func (s *Server) auditFilter(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if s.db == nil || !strings.HasPrefix(r.URL.Path, "/api/") || r.Method == "OPTIONS" || r.URL.Path == "/api/admin-activity" || r.URL.Path == "/api/stripe/webhook" {
			next.ServeHTTP(w, r)
			return
		}
		user, err := s.auth(r)
		if err != nil {
			next.ServeHTTP(w, r)
			return
		}
		page := r.Header.Get("X-Aitok-Page")
		if !adminPagePattern.MatchString(page) {
			page = ""
		}
		resource := safeAuditResource(r.URL.Path)
		ctx := &requestAuditContext{Page: page, Method: r.Method, Resource: resource}
		r = r.WithContext(context.WithValue(r.Context(), requestAuditKey{}, ctx))
		aw := &auditWriter{ResponseWriter: w}
		next.ServeHTTP(aw, r)
		status := aw.status
		if status == 0 {
			status = 200
		}
		mutation := r.Method != "GET" && r.Method != "HEAD"
		// 定时同步与页面轮询不制造操作噪音；用户主动访问、筛选由前端事件覆盖。
		background := r.URL.Path == "/api/proxy-activity" || strings.HasSuffix(r.URL.Path, "/login")
		sensitiveRead := strings.HasSuffix(r.URL.Path, "/export") || (strings.HasPrefix(r.URL.Path, "/api/card-operations/") && r.URL.Query().Get("export") == "1") || auditCardDetails.MatchString(r.URL.Path) || (r.URL.Path == "/api/bank-cards" && r.URL.Query().Get("include_numbers") == "1")
		if ctx.Recorded && status < 400 {
			return
		} // 业务事务内已有完整审计，不重复追加成功记录。
		if (!mutation || background) && !sensitiveRead && !(page != "" && status >= 400) {
			return
		}
		result := "success"
		if status >= 400 {
			result = "failure"
		}
		payload, _ := json.Marshal(map[string]any{"source": "server", "page": page, "method": r.Method, "resource": resource, "status": status, "result": result})
		writeCtx, cancel := context.WithTimeout(context.WithoutCancel(r.Context()), 3*time.Second)
		defer cancel()
		_, err = s.db.ExecContext(writeCtx, `INSERT INTO operation_events(actor_id,entity_type,entity_id,action,request_key,after_data) VALUES($1,'admin_request',0,$2,$3,$4)`, user, r.Method+" "+resource, eventKey(), payload)
		if err != nil {
			log.Print("operation audit write failed")
		}
	})
}

// 未知路径可能夹带凭据，只保留已知资源、数字 ID 和固定动作段。
func safeAuditResource(path string) string {
	known := map[string]bool{"api": true, "accounts": true, "account-groups": true, "addresses": true, "bank-cards": true, "card-operations": true, "orders": true, "packages": true, "users": true, "wallet": true, "topups": true, "audit": true, "proxy-activity": true, "payment-exceptions": true, "notices": true, "order-operators": true, "me": true, "logout": true, "export": true, "import": true, "ledger": true, "access": true, "role": true, "group": true, "login": true, "subscription": true, "browser": true, "browser-session": true, "session": true, "renewal-date": true, "renew": true, "sync": true, "refund": true, "collection-quote": true, "record": true}
	known["billing-address"] = true
	known["payment-card"], known["payment-cards"] = true, true
	known["owner"] = true
	known["desktop-auth"] = true
	known["notes"] = true
	parts := strings.Split(strings.Trim(path, "/"), "/")
	for i, part := range parts {
		if !known[part] && !auditNumericID.MatchString(part) {
			parts[i] = ":id"
		}
	}
	return "/" + strings.Join(parts, "/")
}

var auditNumericID = regexp.MustCompile(`^[0-9]{1,19}$`)

var auditCardDetails = regexp.MustCompile(`^/api/bank-cards/[0-9]+$`)
